service/src/main/java/bio/terra/workspace/service/resource/controlled/flight/clone/dataset/CreateTableCopyJobsStep.java

```diff
@@ -0,0 +1,218 @@
+  }
+
+  /**
+   * Create one BigQuery copy job for each table in the source dataset. Keep a running map from
+   * table ID to job ID as new jobs are created, and only create jobs for tables that aren't in the
+   * map already. Rerun the step after every table is processed so that the map may be persisted
+   * incrementally.
+   *
+   * <p>On retry, create the jobs for any tables that don't have them. Use WRITE_TRUNCATE to avoid
+   * the possibility of duplicate data.
+   */
+  @Override
+  public StepResult doStep(FlightContext flightContext)
+      throws InterruptedException, RetryException {
+    final FlightMap workingMap = flightContext.getWorkingMap();
+    final CloningInstructions effectiveCloningInstructions =
+        flightContext
+            .getInputParameters()
+            .get(ControlledResourceKeys.CLONING_INSTRUCTIONS, CloningInstructions.class);
+    if (CloningInstructions.COPY_RESOURCE != effectiveCloningInstructions) {
+      return StepResult.getStepResultSuccess();
+    }
+
+    // Gather inputs
+    final DatasetCloneInputs sourceInputs = getSourceInputs();
+    workingMap.put(ControlledResourceKeys.SOURCE_CLONE_INPUTS, sourceInputs);
+
+    final DatasetCloneInputs destinationInputs = getDestinationInputs(flightContext);
+    workingMap.put(ControlledResourceKeys.DESTINATION_CLONE_INPUTS, destinationInputs);
+
+    final AuthenticatedUserRequest userRequest =
+        flightContext
+            .getInputParameters()
+            .get(JobMapKeys.AUTH_USER_INFO.getKeyName(), AuthenticatedUserRequest.class);
+    final BigQueryCow bigQueryCow = crlService.createBigQueryCow(userRequest);
+    // TODO(jaycarlton):  remove usage of this client when it's all in CRL PF-942
+    final Bigquery bigQueryClient = crlService.createNakedBigQueryClient(userRequest);
+    try {
+      // Get a list of all tables in the source dataset
+      final TableList sourceTables =
+          bigQueryCow
+              .tables()
+              .list(sourceInputs.getProjectId(), sourceInputs.getDatasetName())
+              .execute();
+      // Start a copy job for each source table
+      final Map<String, String> tableToJobId =
+          Optional.ofNullable(
+                  workingMap.get(
+                      ControlledResourceKeys.TABLE_TO_JOB_ID_MAP,
+                      new TypeReference<Map<String, String>>() {}))
+              .orElseGet(HashMap::new);
+      final List<Tables> tables = sourceTables.getTables();
+      // Find the first table whose ID isn't a key in the map.
+      final Optional<Tables> tableMaybe =
+          tables.stream().filter(t -> !tableToJobId.containsKey(t.getId())).findFirst();
+      if (tableMaybe.isPresent()) {
+        final Tables table = tableMaybe.get();
+        // If the table contains data in its streaming buffer, that data can't be copied yet
+        // https://cloud.google.com/bigquery/streaming-data-into-bigquery#dataavailability
+        final Table tableGetResponse =
```

Comment:
minor: it might make the algorithm more readable if this streaming check was moved to a separate `checkStreaming` method.

Reply:
good idea